### ISSUE TITLE
Card editor: tags

### ISSUE THREAD

**Issue by ackernaut** [ParabolInc/action#1044]
## Enhancement

The card editor supports:

- [x] Pressing `#` triggers autocomplete for custom tags `#private` and `#archive`, and 1 or many `#[objective]` tags
  - [x] Users can type to narrow down tags by autosuggest and hit `enter` to add
  - [x] Users can press `up` and `down` and `enter` to add
  - [x] Users can tap from the list to add
- [x] Inline tags are preceded with the hashtag `#`, are teal/cool/link color, and are bold
- [ ] Tapping a tag reveals an inline menu ( Filter / Remove )
  - [ ] Pressing Filter adds a removable filter to a search/filter dashboard UI area to be designed and shows all cards with that `#[objective]` (including `#private` and `#archive`)
  - [ ] Pressing Remove removes the tag including `#` and text
- [x] Backspacing removes the entire inline tag (including the `#` and text)

**TBD**: Where do `#[objective]`s come from?

**Proposal 1)** They are organically created by team members and create hive organization. When a project card has a new `#topic` it is added to the list. The filter UI in the dashboard and the inline filter menu gets populated from any tags used that are not only found in archived cards.

**Proposal 2)** Objectives are intentionally added on the Team Sheet by a team leader (or members) kind of like Milestones on GitHub. If a leader or team member adds 1 or many objectives to the list, those can then be used as organizing tags that can be added to cards.

**Estimated effort**: 18 points ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))

#WYSIWYG

**Comment by mattkrick** [comment]
I've been playing with a LOT of hashtags lately to dig into this. Here's a 3rd proposal for your consideration:

- As you type in a "hashtag" entity (any word that starts with #) a suggestion mention of filtered suggestions pops down
- At the bottom of the suggestions is a line saying "create a new tag"
- if they create a new tag, it gets stored on the team & organization. Tags that are owned by the team are weighted more & appear higher in the list. (we could pop something up & ask em if they wanna add to just team or just org?)
- if you create a tag without clicking "create a new tag" you can still rock & roll, it's still a tag, _but it won't appear as a suggestion for you or anyone else_. this freedom allows folks to put personality in their content without the big boss seeing #shakeItLikeASaltShaker when he wants to tag something.
- In the settings, (probably team settings) you can click "Manage tags". There, you can CRUD them. The Update is permissions (ie add/remove from the organization/team level)

tl;dr - we wanna find the right balance, which i think is somewhere between twitter and github


Aside from that, it looks _really_ good! only 1 thought: do we really need a menu? If backspacing removes the whole tag, then the only functionality we need to serve is "filter". Could we make it so when you click the tag, it moves it to the filter? We could give the tag some fancy styles or a pointer cursor or anything you like-- it's just a regular component called `Hashtag.js` so possibilities are limitless.

**Comment by ackernaut** [comment]
I like that idea:

- **Just Type a Tag**: #myRandomTag
  - Unorganized, but one can search/filter with it
- **Create a New Tag**: #teamObjectiveTag
  - Adds to a list that populates team sheet, autosuggest, etc.
  - Can search/filter with it
  - Informs reporting/insights?

**Comment by ackernaut** [comment]
@mattkrick are we going to still force all lowercase, no spaces, and no case sensitivity?

**Comment by mattkrick** [comment]
man, that is the question.
OK, let's lay out our options & see what we get:
- case-sensitive, private != Private
- case-insensitive, private == Private
- forced-case, you _will_ type private
- "i know what you mean" case, you type Private, it shows Private on your card, but we still flag it as private
- spaces dis/allowed

I'll give you a couple options that feel right:

Option 1:
- you can use any case you want
- you cannot use spaces
- you cannot create 2 tags that only differ by casing (if you type fOo and foo is a tag, we'll change it to foo for you)
- tags start with a hash

Option 2:
The same as option 1, but we remove the # and let you break up words with spaces instead of forcing you to camelcase.

**Comment by mattkrick** [comment]
manual tests:
- [x] hitting the trigger immediately pops up the suggestion modal
- [x] hitting backspace to remove the trigger removes the suggestion modal
- [x] clicking into a word that begins with the trigger pops up the suggestion modal & the suggestion query includes the whole word
- [x] hitting tab autocompletes with the highlighted suggestion
- [x] hitting enter is same as tab
- [x] hitting up/down arrows move the selection
- [x] clicking on a suggestion autocompletes

**Comment by mattkrick** [comment]
@jordanh @ackernaut above the test comment, could you let me know which option you guys prefer for user-defined tagging?

**Comment by ackernaut** [comment]
🤔 

**Option 1** builds on the meme of hashtags. We assume our market just _gets_ how to do that. We want to see if users are delighted upon realizing they are building organization. We enhance the feature allowing users to identify which hashtags are on the list of team objectives, etc.

**Option 2** makes me think we are designing even further along the lines of named objectives (like GH milestones). I think this requires more cognitive effort (and learning our UX) for users to want to identify outcome cards by linking to named objectives. **Example**: If I link to a Google doc [Our Team Objectives](www.parabol.co) and file this under [Governance](www.parabol.co) does that feel right. Would I know when I am linking versus tagging? What about with [Private](www.parabol.co) and [Archived](www.parabol.co) as hashtagless tags?

**Comment by jordanh** [comment]
Will revisit and open new issues that consider the unfinished portions of this issue. Iceboxing for now.
